client/app/components/dashboards/add-widget-dialog.html

```diff
@@ -4,46 +4,75 @@ <h4 class="modal-title">Add Widget</h4>
 </div>
 <div class="modal-body">
   <p class="btn-group">
-      <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isVisualization()}" ng-click="$ctrl.setType('visualization')">Visualization</button>
-      <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isTextBox()}" ng-click="$ctrl.setType('textbox')">Text Box</button>
+    <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isVisualization}" ng-click="$ctrl.setType('visualization')">Visualization</button>
+    <button type="button" class="btn btn-default" ng-class="{active: $ctrl.isTextBox}" ng-click="$ctrl.setType('textbox')">Text Box</button>
   </p>
 
-  <div ng-show="$ctrl.isTextBox()">
-      <div class="form-group">
-          <textarea class="form-control" ng-model="$ctrl.text" rows="3"></textarea>
+  <div ng-show="$ctrl.isTextBox">
+    <div class="form-group">
+      <textarea class="form-control" ng-model="$ctrl.text" ng-model-options="{ debounce: 200 }" rows="3"></textarea>
+    </div>
+    <div ng-show="$ctrl.text">
+      <strong>Preview:</strong>
+      <p ng-bind-html="$ctrl.text | markdown"></p>
+    </div>
+  </div>
+
+  <div ng-show="$ctrl.isVisualization">
+    <div class="form-group">
+      <input type="text" placeholder="Search a query by name" class="form-control" autofocus="autofocus"
+        ng-if="!$ctrl.selectedQuery" ng-change="$ctrl.searchQueries($ctrl.searchTerm)"
+        ng-model="$ctrl.searchTerm" ng-model-options="{ debounce: 200 }">
+      <div ng-if="$ctrl.selectedQuery" class="p-relative">
+        <input type="text" class="form-control bg-white"
+          ng-value="$ctrl.selectedQuery.name" readonly="readonly">
+        <a href="javascript:void(0)" ng-click="$ctrl.selectQuery(null)"
+          class="d-flex align-items-center justify-content-center"
+          style="position: absolute; right: 1px; top: 1px; bottom: 1px; width: 30px; background: #fff; border-radius: 3px;"
+        ><i class="text-muted fa fa-times"></i></a>
       </div>
-      <div ng-show="$ctrl.text">
-          <strong>Preview:</strong>
-          <p ng-bind-html="$ctrl.text | markdown"></p>
+    </div>
+
+    <div ng-if="!$ctrl.selectedQuery && ($ctrl.searchTerm == '')" class="scrollbox" style="max-height: 50vh">
+      <div class="list-group" ng-if="$ctrl.recentQueries.length > 0">
+        <a href="javascript:void(0)" class="list-group-item"
+          ng-repeat="query in $ctrl.recentQueries" ng-click="$ctrl.selectQuery(query.id)"
+        >
+          <span>{{query.name}}</span>
+          <span class="label label-default" ng-if="query.is_draft">Unpublished</span>
```

Comment:
It not supposed to show unpublished queries here. I guess they show up, because the recents API returns them. You can either apply filtering on the server side or on the client once we have the recents.